tests/integrationtest/r/expression/format.result

```diff
@@ -16,47 +16,56 @@ FORMAT('1.234',1,'xxxxxx')
 SHOW WARNINGS;
 Level	Code	Message
 Warning	1649	Unknown locale: 'xxxxxx'
-SELECT FORMAT(100000.234,4,'nl_BE');
-FORMAT(100000.234,4,'nl_BE')
-100.000,2340
-SELECT FORMAT(100000.234,4,'fr_BE');
-FORMAT(100000.234,4,'fr_BE')
-100 000,2340
-SELECT FORMAT(100000.234,4,'en_IN');
-FORMAT(100000.234,4,'en_IN')
-1,00,000.2340
-SELECT FORMAT(100000.234,4,'de_CH');
-FORMAT(100000.234,4,'de_CH')
-100’000.2340
-SELECT FORMAT(100000.234,4,'no_NO');
-FORMAT(100000.234,4,'no_NO')
-100,000.2340
-SELECT FORMAT(100000.234,4,'bg_BG');
-FORMAT(100000.234,4,'bg_BG')
-100 000,2340
-SELECT FORMAT(100000.234,4,'ar_SA');
-FORMAT(100000.234,4,'ar_SA')
-100,000.2340
-SELECT FORMAT(100000.234,4,'es_MX');
-FORMAT(100000.234,4,'es_MX')
-100,000.2340
-SELECT FORMAT(100000.234,4,'ps_AF');
-FORMAT(100000.234,4,'ps_AF')
-100.000,2340
-SELECT FORMAT(100000.234,4,'kv_RU');
-FORMAT(100000.234,4,'kv_RU')
-100,000.2340
-SELECT FORMAT(100000.234,999999999999999999,'de_CH');
-FORMAT(100000.234,999999999999999999,'de_CH')
-100’000.233999999996740370988845825195
-SELECT FORMAT(100000.234,9999999999999999999,'de_CH');
-FORMAT(100000.234,9999999999999999999,'de_CH')
-100’000
-SELECT FORMAT(100000.234,99999999999999999999,'de_CH');
-FORMAT(100000.234,99999999999999999999,'de_CH')
-100’000.233999999996740370988845825195
+SELECT FORMAT(12345678.9,4,'nl_BE');
+FORMAT(12345678.9,4,'nl_BE')
+12.345.678,9000
+SELECT FORMAT(12345678.9,4,'fr_BE');
+FORMAT(12345678.9,4,'fr_BE')
+12 345 678,9000
+SELECT FORMAT(12345678.9,4,'en_IN');
+FORMAT(12345678.9,4,'en_IN')
+1,23,45,678.9000
+SELECT FORMAT(12345678.9,4,'de_CH');
+FORMAT(12345678.9,4,'de_CH')
+12’345’678.9000
+SELECT FORMAT(12345678.9,4,'no_NO');
+FORMAT(12345678.9,4,'no_NO')
+12,345,678.9000
+SELECT FORMAT(12345678.9,4,'bg_BG');
+FORMAT(12345678.9,4,'bg_BG')
+12 345 678,9000
+SELECT FORMAT(12345678.9,4,'ar_SA');
+FORMAT(12345678.9,4,'ar_SA')
+12,345,678.9000
+SELECT FORMAT(12345678.9,4,'es_MX');
+FORMAT(12345678.9,4,'es_MX')
+12,345,678.9000
+SELECT FORMAT(12345678.9,4,'ps_AF');
+FORMAT(12345678.9,4,'ps_AF')
+12.345.678,9000
+SELECT FORMAT(12345678.9,4,'kv_RU');
+FORMAT(12345678.9,4,'kv_RU')
+12,345,678.9000
+SELECT FORMAT(12345678.9,999999999999999999,'de_CH');
+FORMAT(12345678.9,999999999999999999,'de_CH')
+12’345’678.900000000372529029846191406250
```

Comment:
why will it end with ....372529029846191406250? the input should be a decimal number with exact decimal representation :thinking: 

```
mysql> select FORMAT(12345678.9, 99999999999999999, 'de_CH');
+------------------------------------------------+
| FORMAT(12345678.9, 99999999999999999, 'de_CH') |
+------------------------------------------------+
| 12'345'678.900000000000000000000000000000      |
+------------------------------------------------+
1 row in set (0.00 sec)
```

also why is it separated using `’` U+2019 RIGHT SINGLE QUOTATION MARK rather than `'` U+0027 APOSTROPHE? is this OS-dependent?

Reply:
# Go

```go
package main

import (
	"golang.org/x/text/language"
	"golang.org/x/text/message"
	"golang.org/x/text/number"
)

func main() {
	p := message.NewPrinter(language.MustParse("de_CH"))
	p.Printf("%f", number.Decimal(12345678.9, number.Scale(1)))
}

```
https://go.dev/play/p/7Iv3SA0R-bR

output:
```
12’345’678.9
```

# C with ICU

```c
#include <stdio.h>
#include <unicode/unum.h>
#include <unicode/uloc.h>
#include <unicode/ustring.h>
#include <unicode/ustdio.h>

int main() {
    double number = 12345678.9;
    const char* locale = "de_CH";
    UChar result[100];

    UErrorCode status = U_ZERO_ERROR;
    UNumberFormat* fmt = unum_open(UNUM_DECIMAL, NULL, 0, locale, NULL, &status);
    int32_t resultLength = sizeof(result) / sizeof(result[0]);
    resultLength = unum_formatDouble(fmt, number, result, resultLength, NULL, &status);
    char resultUTF8[100];
    u_strToUTF8(resultUTF8, sizeof(resultUTF8), NULL, result, resultLength, &status);
    printf("%s\n", resultUTF8);
    unum_close(fmt);
    return 0;
}
```

output:
```
12’345’678.9
```

# Python

```python
#!/bin/python3
import locale

locale.setlocale(locale.LC_NUMERIC, "de_CH")
print(locale.format_string("%.2f", 12345678.9, grouping=True))
```

output:
```
12'345'678.90
```

# Conclusion

Depending on the OS, programming language and libraries either `’` or `'` are used. Looks like both are acceptable.

Note that this results in minor differences between TiDB and MySQL (tested with 9.0.1):

| locale | TiDB | MySQL | Difference? |
|---|---|---|---|
|nl_BE|12.345.678,9000|12345678,9000|Yes|
|fr_BE|12 345 678,9000|12345678,9000|Yes|
|en_IN|1,23,45,678.9000|1,23,45,678.9000|No|
|de_CH|12’345’678.9000|12'345'678.9000|Yes|
|no_NO|12,345,678.9000|12.345.678,9000|Yes (big difference)|
|bg_BG|12 345 678,9000|12 345 678,9000|No|
|ar_SA|12,345,678.9000|12345678.9000|Yes|
|es_MX|12,345,678.9000|12,345,678.9000|No|

Reply:
The `no_NO` difference is probably not a real issue as it should have been `nb_NO` or `nn_NO` instead of `no_NO`. See also https://bugzilla.redhat.com/show_bug.cgi?id=532487#c0